src/dwarf/debug_info.rs

```diff
@@ -221,336 +230,140 @@ pub enum AttrValue<'a> {
-            if 2 <= data.len() {
-                let bytes = decode_uhalf(data);
-                let fullsize = bytes as usize + 2;
-                if fullsize <= data.len() {
-                    Some((AttrValue::Bytes(&data[2..fullsize]), fullsize))
-                } else {
-                    None
-                }
-            } else {
-                None
-            }
+            let bytes = data.read_u16()?;
+            let fullsize = bytes as usize + 2;
+            Some((AttrValue::Bytes(data.read_slice(bytes.into())?), fullsize))
         }
         DW_FORM_block4 => {
-            if 2 <= data.len() {
-                let bytes = decode_uword(data);
-                let fullsize = bytes as usize + 2;
-                if fullsize <= data.len() {
-                    Some((AttrValue::Bytes(&data[2..fullsize]), fullsize))
-                } else {
-                    None
-                }
-            } else {
-                None
-            }
-        }
-        DW_FORM_data2 => {
-            if 2 <= data.len() {
-                let v = decode_uhalf(data);
-                Some((AttrValue::Unsigned(v as u64), 2))
-            } else {
-                None
-            }
-        }
-        DW_FORM_data4 => {
-            if 4 <= data.len() {
-                let v = decode_uword(data);
-                Some((AttrValue::Unsigned(v as u64), 4))
-            } else {
-                None
-            }
-        }
-        DW_FORM_data8 => {
-            if 8 <= data.len() {
-                let v = decode_udword(data);
-                Some((AttrValue::Unsigned(v), 8))
-            } else {
-                None
-            }
+            let bytes = data.read_u32()?;
+            let fullsize = bytes as usize + 2;
+            Some((AttrValue::Bytes(data.read_slice(bytes as usize)?), fullsize))
         }
+        DW_FORM_data2 => Some((AttrValue::Unsigned(data.read_u16()?.into()), 2)),
+        DW_FORM_data4 => Some((AttrValue::Unsigned(data.read_u32()?.into()), 4)),
+        DW_FORM_data8 => Some((AttrValue::Unsigned(data.read_u64()?), 8)),
         DW_FORM_string => {
-            let s = extract_string(data, 0)?;
-            if (s.len() + 1) <= data.len() {
-                Some((AttrValue::String(s), s.len() + 1))
-            } else {
-                None
-            }
+            let string = data.read_cstr()?;
+            let len = string.to_bytes_with_nul().len();
+            Some((AttrValue::String(string.to_str().ok()?), len))
         }
         DW_FORM_block => {
-            let (sz, bytes) = decode_leb128_128(data)?;
+            let (sz, bytes) = data.read_u128_leb128()?;
             let fullsz = sz as usize + bytes as usize;
-            if fullsz <= data.len() {
-                Some((AttrValue::Bytes(&data[bytes as usize..fullsz]), fullsz))
-            } else {
-                None
-            }
+            Some((AttrValue::Bytes(data.read_slice(bytes.into())?), fullsz))
```

Comment:
so `sz` is value of leb128-encoded size prefix, while `bytes` is how many bytes that leb128 encoding took, so should we `read_slice(sz)` then?

This confusion seems to come from inconsistent naming approach. Here `sz` is length of data, while everywhere else `bytes` is that.

Reply:
Yes, should be `sz`. Good catch! Will see if we can name things more uniformly.

Reply:
Ah, so I caught that earlier and it is already fixed. But I merged the fix into the wrong commit. So if you check the next one there we use the correct size, actually: https://github.com/libbpf/blazesym/pull/56/commits/ffb567eedcb9d11ba214a0c9ff52d511f9d30a45#diff-90ecd21dadac50f04b51aad6a191e1e07361d83bbedd11fb19c9ac2bf2943928R247